src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpCookieHelper
+    {
+        private const string CookieHeaderNameWithColon = HttpKnownHeaderNames.Cookie + ":";
+
+        public static void AddResponseCookiesToContainer(WinHttpRequestState state)
+        {
+            HttpRequestMessage request = state.RequestMessage;
+            SafeWinHttpHandle requestHandle = state.RequestHandle;
+            CookieContainer cookieContainer = state.Handler.CookieContainer;
+            
+            Debug.Assert(state.Handler.CookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer);
+            Debug.Assert(cookieContainer != null);
+
+            // Get 'Set-Cookie' headers from response
+            List<string> cookieHeaders =
+                WinHttpResponseParser.GetResponseHeaders(requestHandle, Interop.WinHttp.WINHTTP_QUERY_SET_COOKIE);
+            WinHttpTraceHelper.Trace("WINHTTP_QUERY_SET_COOKIE");
+            
+            foreach (string cookieHeader in cookieHeaders)
+            {
+                WinHttpTraceHelper.Trace(cookieHeader);
+                try
+                {
+                    cookieContainer.SetCookies(request.RequestUri, cookieHeader);
```

Comment:
request.RequestUri is updated during auto redirection in the WinHttpStatusCallback.  So, it will have been updated already before this code above runs in the case where this method is called after a terminal response.  

In the case where it is called during the actual redirection callback, then it uses the current URI (not the new one yet) because it needs to put cookies into the container based on the the current request being processed.

So, in both cases, the correct URI is used.